### PR TITLE
rand: configuration based DRBG selection.

### PR DESCRIPTION
**Pull request by paulidale**
##### Checklist

- [x] documentation is added or updated
- [ ] tests are added or updated

### CODE REVIEW

File: crypto/rand/rand_lib.c [hunk 1]
```diff
@@ -565,3 +595,74 @@ EVP_RAND_CTX *RAND_get0_private(OPENSSL_CTX *ctx)
     }
     return rand;
 }
+
+#ifndef FIPS_MODULE
+static int random_set_string(char **p, const char *s)
+{
+    char *d = strdup(s);
```
Comment (mattcaswell) [inline crypto/rand/rand_lib.c:602]: OPENSSL_strdup?

**Review by mattcaswell — CHANGES_REQUESTED** [review]
doc/man5/config.pod could do with an update to document the new parameters.

This also needs a test

**Comment by paulidale** [comment]
How do we test configuration settings???

**Comment by mattcaswell** [comment]
>  How do we test configuration settings???

Probably we will need to reach into the internals and check we got the right type of EVP_RAND instances created.

File: crypto/rand/rand_lib.c (hunk 1 above)
Comment (paulidale) [reply crypto/rand/rand_lib.c:602]: Good point, I often miss this one :(

**Comment by paulidale** [comment]
Added some words to `config.pod`.  I'll try to figure out the test tomorrow.

I'm half tempted to change the `random=name` to `drbg=name` in the random section.

File: doc/man5/config.pod [hunk 2]
```diff
@@ -326,69 +330,41 @@ key types, as shown here:
  RSA.Certificate = server-rsa.pem
  ECDSA.Certificate = server-ecdsa.pem
 
-=head2 Engine Configuration
```
Comment (mattcaswell) [inline doc/man5/config.pod:329]: Why delete the Engine section?

File: doc/man5/config.pod [hunk 3]
```diff
@@ -326,69 +330,41 @@ key types, as shown here:
...
+Within the random section, the following names have meaning:
 
 =over 4
 
-=item B<engine_id>
+=item B<random>
 
-This is used to specify an alternate name, overriding the default name
-specified in the list of engines. If present, it must be first.
+This is used to specify the random bit generator.
```
Comment (mattcaswell) [inline doc/man5/config.pod:344]: Could we list the internally supported DRBG names?

**Review by mattcaswell — CHANGES_REQUESTED** [review]

**Comment by mattcaswell** [comment]
> I'm half tempted to change the random=name to drbg=name in the random section.

No strong view on this.

File: doc/man5/config.pod (hunk 2 above)
Comment (paulidale) [reply doc/man5/config.pod:329]: Editing error :(  It's late.

Comment (paulidale) [reply doc/man5/config.pod]: Should be back now.

File: doc/man5/config.pod (hunk 3 above)
Comment (paulidale) [reply doc/man5/config.pod:407]: I've added a list.  I also included DRBG names in the _cipher_ and _digest_ names.

File: crypto/rand/rand_lib.c [hunk 4]
```diff
@@ -565,3 +595,74 @@ EVP_RAND_CTX *RAND_get0_private(OPENSSL_CTX *ctx)
...
+    RAND_GLOBAL *dgbl = rand_get_global(cnf->libctx);
+    int i, r = 1;
+
+    OSSL_TRACE1(CONF, "Loading random module: section %s\n",
+                CONF_imodule_get_value(md));
+
+    /* Value is a section containing RANDOM configuration */
+    elist = NCONF_get_section(cnf, CONF_imodule_get_value(md));
+
+    if (!elist) {
```
Comment (richsalz) [inline crypto/rand/rand_lib.c:629]: explicit test against NULL and extra blank line at 628.

File: doc/man5/config.pod [hunk 5]
```diff
@@ -389,6 +393,53 @@ For example:
...
+
+=item B<HMAC-DRBG>
+
+=back
+
+=item B<cipher>
+
+This specifies what cipher a B<CTR-DRBG> random bit generator will use.
+Other random bit generators ignore this name.
+The default value is B<AES-256-CTR> adn
```
Comment (richsalz) [inline doc/man5/config.pod:429]: adn?

File: crypto/rand/rand_lib.c (hunk 4 above)
Comment (paulidale) [reply crypto/rand/rand_lib.c]: Argh, I didn't notice this when copy/pasting.  Fixed now.

File: doc/man5/config.pod (hunk 5 above)
Comment (paulidale) [reply doc/man5/config.pod]: Removed and replaced by a full stop.

**Comment by paulidale** [comment]
Test case added, ready for review now.

**Review by mattcaswell — APPROVED** [review]
LGTM

**Comment by paulidale** [comment]
Merged to master (an hour or so early).

**Pull request closed without merging**